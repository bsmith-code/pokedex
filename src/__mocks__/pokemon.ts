import {
  IEvolutionChain,
  INamedApiResourceList,
  IPokemon,
  IPokemonAbility,
  IPokemonMove,
  IPokemonMoveVersion,
  IPokemonSpecies,
  IPokemonType
} from 'pokeapi-typescript'

const mockApiUrl = 'https://www.mockPokeApi.co/api/v2'
export const mockPokemonList: INamedApiResourceList<{
  name: string
  url: string
}> = {
  count: 100,
  next: '',
  previous: '',
  results: Array.from({ length: 100 }).map((_, idx) => ({
    name: `pokemon-${idx}`,
    url: `${mockApiUrl}/pokemon/${idx}`
  }))
}

export const mockMoves: IPokemonMove[] = Array.from({ length: 20 }).map(
  (_, idx) => ({
    move: {
      name: `mock-move-${idx}`,
      url: `${mockApiUrl}/move/${idx}`
    },
    version_group_details: [] as IPokemonMoveVersion[]
  })
)

export const mockAbilities: IPokemonAbility[] = Array.from({ length: 20 }).map(
  (_, idx) => ({
    ability: {
      name: `mock-ability-${idx}`,
      url: `${mockApiUrl}/ability/${idx}`
    },
    is_hidden: true,
    slot: idx
  })
)

export const mockTypes: IPokemonType[] = Array.from({ length: 20 }).map(
  (_, idx) => ({
    slot: idx,
    type: {
      name: `mock-type-${idx}`,
      url: `${mockApiUrl}/type/${idx}`
    }
  })
)

export const mockPokemonDetails: IPokemon = {
  abilities: mockAbilities,
  base_experience: 20,
  forms: [],
  game_indices: [],
  height: 20,
  held_items: [],
  id: 1,
  is_default: true,
  location_area_encounters: '',
  moves: mockMoves,
  name: 'mock-pokemon',
  order: 1,
  species: {
    name: 'Mock Species',
    url: `${mockApiUrl}/pokemon-species/3`
  },
  sprites: {
    front_default: `${mockApiUrl}/sprites/pokemon/front/1`,
    front_shiny: `${mockApiUrl}/sprites/pokemon/shiny/1`,
    front_female: `${mockApiUrl}/sprites/pokemon/front-female/1`,
    front_shiny_female: `${mockApiUrl}/sprites/pokemon/shiny-female/1`,
    back_default: `${mockApiUrl}/sprites/pokemon/back/1`,
    back_shiny: `${mockApiUrl}/sprites/pokemon/back-shiny/1`,
    back_female: `${mockApiUrl}/sprites/pokemon/back-female/1`,
    back_shiny_female: `${mockApiUrl}/sprites/pokemon/back-shiny-female/1`,
    other: {
      'official-artwork': {
        front_default:
          'https://mockPokeApi.com/sprites/pokemon/other/official-artwork/1'
      }
    }
  },
  stats: [],
  types: mockTypes,
  weight: 1000
}

export const mockPokemonSpecies: IPokemonSpecies = {
  id: 3,
  name: 'mock-species',
  order: 1,
  gender_rate: 1,
  capture_rate: 20,
  base_happiness: 1000,
  is_baby: false,
  hatch_counter: 50,
  has_gender_differences: true,
  forms_switchable: true,
  growth_rate: {
    name: 'mock-growth-rate',
    url: `${mockApiUrl}/growth-rate/5`
  },
  pokedex_numbers: [],
  egg_groups: [],
  color: {
    name: 'mock-color',
    url: `${mockApiUrl}/pokemon-color/5`
  },
  shape: {
    name: 'mock-shape',
    url: `${mockApiUrl}/pokemon-shape/8`
  },
  evolves_from_species: {
    name: 'mock-species',
    url: `${mockApiUrl}/pokemon-species/1`
  },
  evolution_chain: {
    url: `${mockApiUrl}/pokemon-species/10`
  },
  habitat: {
    name: 'mock-habitat',
    url: `${mockApiUrl}/pokemon-habitat/30`
  },
  generation: {
    name: 'mock-generation',
    url: `${mockApiUrl}/pokemon-generation/4`
  },
  names: [],
  pal_park_encounters: [],
  flavor_text_entries: [],
  form_descriptions: [],
  genera: [],
  varieties: []
}

export const mockPokemonEvolutions: IEvolutionChain = {
  baby_trigger_item: {
    name: 'mock-item',
    url: `${mockApiUrl}/pokemon-item/10`
  },
  id: 1,
  chain: {
    evolution_details: [],
    evolves_to: [
      {
        evolution_details: [],
        evolves_to: [
          {
            evolution_details: [],
            evolves_to: [],
            is_baby: false,
            species: {
              name: 'mock-species-3',
              url: `${mockApiUrl}/pokemon-species/3`
            }
          }
        ],
        is_baby: false,
        species: {
          name: 'mock-species-2',
          url: `${mockApiUrl}/pokemon-species/2`
        }
      }
    ],
    is_baby: false,
    species: {
      name: 'mock-species',
      url: `${mockApiUrl}/pokemon-species/1`
    }
  }
}
