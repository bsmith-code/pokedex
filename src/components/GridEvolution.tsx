import { IEvolutionChain } from 'pokeapi-typescript'

import { Box, Grid, ListItemButton, styled, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { flattenChain, getIdFromUrl, getPokemonSprite } from 'utils'

interface IProps {
  isFetching: boolean
  evolutions?: IEvolutionChain
}

const StyledGridItem = styled(ListItemButton)(({ theme }) => ({
  flexDirection: 'column',
  borderRadius: '4px',
  color: '#fff',
  padding: theme.spacing(4),
  backgroundColor: theme.palette.secondary.main,
  '&:hover': {
    backgroundColor: '#6704c9'
  }
}))

export const GridEvolution = ({ evolutions, isFetching }: IProps) => {
  const { handleChangePokemon } = usePokemonContext()
  const flattenedChain = flattenChain(evolutions?.chain)

  return !isFetching ? (
    <Box>
      <Typography variant="h6" sx={{ mb: 2 }}>
        Evolutions
      </Typography>
      {flattenedChain?.length ? (
        <Grid container spacing={2}>
          {flattenedChain.map(({ url, name }) => {
            const id = getIdFromUrl(url)
            return (
              <Grid item xs={12} sm={6} md={6} lg={4} key={`evolution-${name}`}>
                <StyledGridItem onClick={() => handleChangePokemon(name)}>
                  <img
                    alt={name}
                    src={getPokemonSprite(id)}
                    width="120px"
                    height="120px"
                    style={{ objectFit: 'contain' }}
                  />
                  <Typography variant="subtitle2">{name}</Typography>
                </StyledGridItem>
              </Grid>
            )
          })}
        </Grid>
      ) : (
        <Typography>No available evolutions</Typography>
      )}
    </Box>
  ) : (
    <Typography>Loading...</Typography>
  )
}
