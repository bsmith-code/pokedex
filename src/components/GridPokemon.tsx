import {
  Box,
  Grid,
  ListItemButton,
  Pagination,
  styled,
  Typography
} from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { getIdFromUrl, getPokemonSprite } from 'utils'

import {
  TESTID_GRID_ERROR,
  TESTID_GRID_LOADING,
  TESTID_GRID_POKEMON
} from 'constants/testIds'

const StyledGridItem = styled(ListItemButton)`
  flex-direction: column;
  border-radius: 4px;
  border: 1px solid #efefef;
`

export const GridPokemon = () => {
  const {
    pokemon,
    currentPage,
    paginatedPokemon,
    isErrorPokemon,
    isFetchingPokemon,
    handleChangePage,
    handleChangePokemon
  } = usePokemonContext()

  if (isFetchingPokemon) {
    return <div data-testid={TESTID_GRID_LOADING}>Loading...</div>
  }

  if (!pokemon.length || isErrorPokemon) {
    return <div data-testid={TESTID_GRID_ERROR}>Pokemon unavailable.</div>
  }

  return (
    <Box mt={4} data-testid={TESTID_GRID_POKEMON}>
      <Grid container spacing={2}>
        {paginatedPokemon.map(({ url, name }) => {
          const imageSrc = getPokemonSprite(getIdFromUrl(url))

          return (
            <Grid
              item
              xs={12}
              sm={6}
              md={4}
              lg={3}
              key={`pokemon-grid-${name}`}
            >
              <StyledGridItem onClick={() => handleChangePokemon(name)}>
                <img
                  alt={name}
                  src={imageSrc}
                  width="120px"
                  height="120px"
                  style={{ objectFit: 'contain' }}
                />
                <Typography>{name}</Typography>
              </StyledGridItem>
            </Grid>
          )
        })}
      </Grid>
      <Box mt={6} display="flex" justifyContent="center">
        <Pagination
          page={currentPage}
          count={Math.floor(pokemon.length / 20)}
          onChange={(_, page) => handleChangePage(page)}
        />
      </Box>
    </Box>
  )
}
