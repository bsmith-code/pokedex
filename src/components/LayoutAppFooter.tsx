import { Toolbar, Typography } from '@mui/material'

const LayoutAppFooter = () => {
  const year = new Date().getFullYear()

  return (
    <Toolbar
      variant="dense"
      component="footer"
      sx={{ bgcolor: 'primary.main', justifyContent: 'center' }}
    >
      <Typography color="#fff">
        &copy; Brian Matthew Smith, {year}. All Rights Reserved.
      </Typography>
    </Toolbar>
  )
}

export default LayoutAppFooter
