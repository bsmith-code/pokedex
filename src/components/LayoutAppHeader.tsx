import { useNavigate } from 'react-router-dom'

import { IconButton, Toolbar } from '@mui/material'

const LayoutAppHeader = () => {
  const navigate = useNavigate()

  const handleLogoClick = () => {
    navigate('/')
  }

  return (
    <Toolbar component="header" sx={{ bgcolor: 'primary.main' }}>
      <IconButton onClick={handleLogoClick}>
        <img
          alt="PrizePicks"
          src="/assets/logo.svg"
          width="140px"
          height="auto"
        />
      </IconButton>
    </Toolbar>
  )
}

export default LayoutAppHeader
