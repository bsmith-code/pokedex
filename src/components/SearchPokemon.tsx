import { INamedApiResource } from 'pokeapi-typescript'

import {
  Autocomplete,
  createFilterOptions,
  ListItemButton,
  TextField
} from '@mui/material'

import { useChangePokemon } from 'hooks/useChangePokemon'

interface IProps {
  pokemon: INamedApiResource<void>[]
}

export const SearchPokemon = ({ pokemon }: IProps) => {
  const { handleChangePokemon } = useChangePokemon()

  return (
    <Autocomplete
      filterOptions={createFilterOptions({ limit: 20 })}
      onChange={(_, value) => handleChangePokemon(value?.name ?? '')}
      options={pokemon}
      renderInput={params => <TextField {...params} label="Search Pokemon" />}
      getOptionLabel={({ name }) => name}
      renderOption={(props, { name }) => (
        <ListItemButton
          {...props}
          component="li"
          key={`pokemon-list-${name}`}
          data-testid={`pokemon-${name}`}
        >
          {name}
        </ListItemButton>
      )}
    />
  )
}
