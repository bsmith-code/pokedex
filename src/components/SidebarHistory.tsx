import { Box, List, ListItemButton, Paper, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { TESTID_SEARCH_EMPTY, TESTID_SEARCH_HISTORY } from 'constants/testIds'

export const SidebarHistory = () => {
  const { searchHistory, handleChangePokemon } = usePokemonContext()

  return (
    <Box
      px={4}
      py={3}
      flexShrink={1}
      flexBasis={300}
      component={Paper}
      bgcolor="grey.200"
      sx={{ '@media (max-width: 767px)': { order: 2, flexGrow: 1 } }}
    >
      <Typography variant="subtitle2">Search History</Typography>

      {searchHistory.length ? (
        <List data-testid={TESTID_SEARCH_HISTORY}>
          {searchHistory.map(name => (
            <ListItemButton
              component="li"
              key={`search-${name}`}
              onClick={() => handleChangePokemon(name)}
            >
              {name}
            </ListItemButton>
          ))}
        </List>
      ) : (
        <Typography data-testid={TESTID_SEARCH_EMPTY} sx={{ mt: 2 }}>
          No previous searches.
        </Typography>
      )}
    </Box>
  )
}
