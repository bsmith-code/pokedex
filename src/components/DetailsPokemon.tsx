import { useNavigate } from 'react-router-dom'

import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import {
  Box,
  Card,
  CardContent,
  CardHeader,
  IconButton,
  Typography
} from '@mui/material'

import { usePokemonDetails } from 'hooks/usePokemonDetails'

import { GridEvolution } from 'components/GridEvolution'

import {
  TESTID_DETAILS_ERROR,
  TESTID_DETAILS_LOADING,
  TESTID_DETAILS_POKEMON,
  TESTID_DETAILS_RETURN
} from 'constants/testIds'

export const DetailsPokemon = () => {
  const {
    name,
    details,
    sprites,
    evolutions,
    speciesName,
    isErrorDetails,
    isFetchingDetails,
    isFetchingEvolutions
  } = usePokemonDetails()
  const navigate = useNavigate()

  const handleGoBack = () => {
    navigate('/')
  }

  if (isFetchingDetails) {
    return <div data-testid={TESTID_DETAILS_LOADING}>Loading...</div>
  }

  if (isErrorDetails) {
    return <div data-testid={TESTID_DETAILS_ERROR}>Invalid Pokemon.</div>
  }

  return (
    <>
      <Box display="flex" mt={2} data-testid={TESTID_DETAILS_POKEMON}>
        <Box flexShrink={0} position="relative">
          <IconButton
            onClick={handleGoBack}
            data-testid={TESTID_DETAILS_RETURN}
            sx={{ position: 'absolute', top: 0, left: 0 }}
          >
            <ArrowBackIcon />
          </IconButton>
          <img
            alt={name}
            src={sprites?.other?.['official-artwork']?.front_default ?? ''}
          />
        </Box>
        <Box p={2} flexGrow={1}>
          <Box mb={2}>
            <Typography component="h1" variant="h4">
              {name}
            </Typography>
            <Typography fontSize={14} color="grey.700" sx={{ mt: 1 }}>
              species: {speciesName}
            </Typography>
          </Box>
          {details.map(({ title, content }) => (
            <Card
              key={`details-${title}`}
              sx={{ bgcolor: 'secondary.main', color: 'common.white', mb: 2 }}
            >
              <CardHeader title={title} />
              <CardContent>{content}</CardContent>
            </Card>
          ))}
        </Box>
      </Box>
      <GridEvolution
        evolutions={evolutions}
        isFetching={isFetchingEvolutions}
      />
    </>
  )
}
