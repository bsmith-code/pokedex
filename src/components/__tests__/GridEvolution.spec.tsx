import { BrowserRouter, Route, Routes } from 'react-router-dom'
import { mockPokemonEvolutions } from '__mocks__/pokemon'
import { fireEvent, screen, waitFor, within } from '@testing-library/react'

import { PokemonContextProvider } from 'context/PokemonContext'

import { GridEvolution } from 'components/GridEvolution'

import { TESTID_GRID_EVOLUTIONS, TESTID_GRID_LOADING } from 'constants/testIds'

const defaultProps = {
  isFetching: false,
  evolutions: mockPokemonEvolutions
}

const MockComponent = (overrideProps: Partial<typeof defaultProps>) => (
  <BrowserRouter>
    <Routes>
      <Route
        index
        path="/pokemon?/:pokemonName?"
        element={
          <PokemonContextProvider>
            <GridEvolution {...defaultProps} {...overrideProps} />
          </PokemonContextProvider>
        }
      />
    </Routes>
  </BrowserRouter>
)

describe('GridEvolution', () => {
  it('renders correctly while loading', async () => {
    const { container } = customRender(<MockComponent isFetching />)

    await waitFor(() => {
      expect(screen.getByTestId(TESTID_GRID_LOADING)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  it('renders correctly with data', async () => {
    const { container } = customRender(<MockComponent />)

    await waitFor(() => {
      expect(screen.getByTestId(TESTID_GRID_EVOLUTIONS)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  it('correctly changes pokemon', async () => {
    customRender(<MockComponent />)

    const evolutions = screen.getByTestId(TESTID_GRID_EVOLUTIONS)
    const evolutionBtn = within(evolutions).getAllByRole('button')[0]
    fireEvent.click(evolutionBtn)

    await waitFor(() => {
      expect(window.location.pathname).toBe(
        `/pokemon/${evolutionBtn.textContent ?? ''}`
      )
    })
  })
})
