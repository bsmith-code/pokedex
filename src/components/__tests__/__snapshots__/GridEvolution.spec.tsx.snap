// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridEvolution renders correctly while loading 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body1 emotion-0"
    data-testid="grid-loading"
  >
    Loading...
  </p>
</div>
`;

exports[`GridEvolution renders correctly with data 1`] = `
<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="grid-evolutions"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 emotion-1"
    >
      Evolutions
    </h6>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 emotion-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-4 emotion-3"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-4"
          role="button"
          tabindex="0"
        >
          <img
            alt="mock-species"
            height="120px"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png"
            style="object-fit: contain;"
            width="120px"
          />
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 emotion-5"
          >
            mock-species
          </h6>
          <span
            class="MuiTouchRipple-root emotion-6"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-4 emotion-3"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-4"
          role="button"
          tabindex="0"
        >
          <img
            alt="mock-species-2"
            height="120px"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png"
            style="object-fit: contain;"
            width="120px"
          />
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 emotion-5"
          >
            mock-species-2
          </h6>
          <span
            class="MuiTouchRipple-root emotion-6"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-4 emotion-3"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters emotion-4"
          role="button"
          tabindex="0"
        >
          <img
            alt="mock-species-3"
            height="120px"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/3.png"
            style="object-fit: contain;"
            width="120px"
          />
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 emotion-5"
          >
            mock-species-3
          </h6>
          <span
            class="MuiTouchRipple-root emotion-6"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
